src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -123,40 +166,91 @@ private async Task ProcessQueueAsync()
             Solution? lastMutatedSolution = null;
-            var queueToken = _cancelSource.Token;
 
-            while (!queueToken.IsCancellationRequested)
+            try
             {
-                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+                while (!_cancelSource.IsCancellationRequested)
+                {
+                    var work = await _queue.DequeueAsync().ConfigureAwait(false);
 
-                // Create a linked cancellation token to cancel any requests in progress when this shuts down
-                var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(queueToken, work.CancellationToken).Token;
+                    // Create a linked cancellation token to cancel any requests in progress when this shuts down
+                    var cancellationToken = CancellationTokenSource.CreateLinkedTokenSource(_cancelSource.Token, work.CancellationToken).Token;
 
-                // The "current" solution can be updated by non-LSP actions, so we need it, but we also need
-                // to merge in the changes from any mutations that have been applied to open documents
-                var (document, solution) = _solutionProvider.GetDocumentAndSolution(work.TextDocument, work.ClientName);
-                solution = MergeChanges(solution, lastMutatedSolution);
+                    // The "current" solution can be updated by non-LSP actions, so we need it, but we also need
+                    // to merge in the changes from any mutations that have been applied to open documents
+                    var (document, solution) = _solutionProvider.GetDocumentAndSolution(work.TextDocument, work.ClientName);
+                    solution = MergeChanges(solution, lastMutatedSolution);
 
-                Solution? mutatedSolution = null;
-                var context = new RequestContext(solution, work.ClientCapabilities, work.ClientName, document, s => mutatedSolution = s);
+                    Solution? mutatedSolution = null;
+                    var context = new RequestContext(solution, work.ClientCapabilities, work.ClientName, document, s => mutatedSolution = s);
 
-                if (work.MutatesSolutionState)
-                {
-                    // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
-                    var ranToCompletion = await work.CallbackAsync(context, cancellationToken).ConfigureAwait(false);
+                    if (work.MutatesSolutionState)
+                    {
+                        // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                        var ranToCompletion = false;
+                        try
+                        {
+                            ranToCompletion = await work.CallbackAsync(context, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
```

Comment:
So now that we're having to do these catches _anyways_, does do we really need the ranToCompletion boolean as well? If instead work.CallbackAsync just returns the same task from the TaskCompletionSource it returned to the original caller, then you can just simplify this to:

```
try
{
   await work.CallbackAsync(....);
   lastMutatedSolution = mutatedSolution ?? lastMutatedSolution;
}
catch (OperationCanceledException)
{
}
catch (Exception e)
{
    OnRequestServerShutdown($"An error occured processing a mutating request and the solution is in an invalid state. Check LSP client logs for any error information.");
}
```

Which also means you now have the exception in hand if you want to pass a message back to the OnRequestServerShutdown rather than eating the exception if the logs are unavailable.

Reply:
This could work, but I don't think I'm a big fan because the Callback will have to catch and rethrow the exceptions, and I think it will generally be harder to understand what is going on. You're right it would make for a better log message, which might be useful.